lib/node_modules/@stdlib/math/base/special/log2/src/main.c

```diff
@@ -0,0 +1,220 @@
+
+
+// BEGIN: polyval_p
+
+/**
+* Evaluates a polynomial.
+*
+* ## Notes
+*
+* -   The implementation uses [Horner's rule][horners-method] for efficient computation.
+*
+* [horners-method]: https://en.wikipedia.org/wiki/Horner%27s_method
+*
+* @param x    value at which to evaluate the polynomial
+* @return     evaluated polynomial
+*/
+static double polyval_p( const double x ) {
+	return 0.3999999999940942 + (x * (0.22222198432149784 + (x * 0.15313837699209373)));
+}
+
+// END: polyval_p// BEGIN: polyval_q
+
+/**
+* Evaluates a polynomial.
+*
+* ## Notes
+*
+* -   The implementation uses [Horner's rule][horners-method] for efficient computation.
+*
+* [horners-method]: https://en.wikipedia.org/wiki/Horner%27s_method
+*
+* @param x    value at which to evaluate the polynomial
+* @return     evaluated polynomial
+*/
+static double polyval_q( const double x ) {
+	return 0.6666666666666735 + (x * (0.2857142874366239 + (x * (0.1818357216161805 + (x * 0.14798198605116586)))));
+}
+
+// END: polyval_q
+
+// MAIN //
+
+/**
+* Return log(x) - (x-1) for x in ~[sqrt(2)/2, sqrt(2)].
+*
+* @param x    input value
+* @return     output value
+*/
+double klog( const double x ) {
+	double hfsq;
+	uint32_t hx;
+	int32_t i;
+	int32_t j;
+	double t1;
+	double t2;
+	double f;
+	double R;
+	double s;
+	double z;
+	double w;
+
+	stdlib_base_float64_get_high_word( x, &hx );
+	f = x - 1.0;
+	if ( ( HIGH_SIGNIFICAND_MASK & ( 2 + hx ) ) < 3 ) {
+		// Case: -2**-20 <= f < 2**-20
+		if ( f == 0.0 ) {
+			return 0.0;
+		}
+		return f * f * ( ( ONE_THIRD*f )- 0.5 );
+	}
+	s = f / ( 2.0 + f );
+	z = s * s;
+	hx &= HIGH_SIGNIFICAND_MASK;
+	i = ( hx - 0x6147a ); 
+	w = z * z;
+	j = ( 0x6b851 - hx );
+	t1 = w * polyval_p( w );
+	t2 = z * polyval_q( w );
+	i |= j;
+	R = t2 + t1;
+	if ( i > 0 ) {
+		hfsq = 0.5 * f * f;
+		return ( s * (hfsq+R) ) - hfsq;
+	}
+	return s * (R-f);
+}
+
+/**
+* Evaluates the binary logarithm (base two).
+*
+* @param x    input value
+* @return     output value
+*/
+double stdlib_base_log2( const double x ) {
+	uint32_t hx;
+	uint32_t lx;
+	double hi;
+	int32_t i;
+	int32_t k;
+	double lo;
+	double xc;
+	double f;
+
+	xc = x;
+	if ( stdlib_base_is_nan( x ) || x < 0.0 ) {
+		return 0.0 / 0.0; // NaN
+	}
+	stdlib_base_float64_to_words( xc, &hx, &lx );
+	k = 0; 
+	if ( hx < HIGH_MIN_NORMAL_EXP ) {
+		// Case: x < 2**-1022
+		if ( ( ( hx & STDLIB_CONSTANT_FLOAT64_HIGH_WORD_ABS_MASK ) | lx ) == 0 ) {
+			return STDLIB_CONSTANT_FLOAT64_NINF;
+		}
+		k -= 54;
+
+		// Subnormal number, scale up x:
+		xc *= TWO54;
+		stdlib_base_float64_get_high_word( xc, &hx );
+	}
+	if ( hx >= HIGH_MAX_NORMAL_EXP ) {
+		return xc + xc;
+	}
+	k += ( ( hx>>20 ) - STDLIB_CONSTANT_FLOAT64_EXPONENT_BIAS );
+	hx &= HIGH_SIGNIFICAND_MASK;
+	i = ( ( hx + 0x95f64 ) & 0x100000 );
+
+	// Normalize x or x/2...
+	stdlib_base_float64_set_high_word( hx|( i^HIGH_BIASED_EXP_0 ), &xc );
+	k += (i>>20);
+	f = klog( xc );
+	xc -= 1;
+	stdlib_base_float64_set_low_word( xc, &hi );
```

Comment:
This does not look correct. The first argument to should be a `uint32_t`, but that is not what is being passed here, as `xc` is a double.

Reply:
What we are wanting here is to create a double with the same high word as `xc` and a low word of `0`. So,

```c
hi = xc;
stdlib_base_float64_set_low_word( 0, &hi );
```

should work.

Reply:
Perfect, I took it differently.